src/main/java/org/broadinstitute/hellbender/tools/HaplotypeCallerSpark.java

```diff
@@ -0,0 +1,329 @@
+
+        if ( !reference.isCompatibleWithSparkBroadcast()){
+            throw new UserException.Require2BitReferenceForBroadcast();
+        }
+
+        final Broadcast<ReferenceMultiSource> referenceBroadcast = ctx.broadcast(reference);
+        final Broadcast<HaplotypeCallerArgumentCollection> hcArgsBroadcast = ctx.broadcast(hcArgs);
+        final OverlapDetector<ReadShard.ShardBoundary> overlaps = getShardBoundaryOverlapDetector(header, intervals, shardingArgs.readShardSize, shardingArgs.readShardPadding);
+        final Broadcast<OverlapDetector<ReadShard.ShardBoundary>> shardBoundariesBroadcast = ctx.broadcast(overlaps);
+
+        final JavaRDD<ReadShard> readShards = createReadShards(shardBoundariesBroadcast, reads);
+
+        final JavaRDD<Tuple2<AssemblyRegion, SimpleInterval>> assemblyRegions = readShards
+                .flatMap(shardsToAssemblyRegions(authHolder, referenceBroadcast, hcArgsBroadcast, shardingArgs, header));
+
+        return assemblyRegions.flatMap(callVariantsFromAssemblyRegions(authHolder, header, referenceBroadcast, hcArgsBroadcast));
+    }
+
+    /**
+     * Call variants from Tuples of AssemblyRegion and Simple Interval
+     * The interval should be the non-padded shard boundary for the shard that the corresponding AssemblyRegion was
+     * created in, it's used to eliminate redundant variant calls at the edge of shard boundaries.
+     */
+    private static FlatMapFunction<Tuple2<AssemblyRegion, SimpleInterval>, VariantContext> callVariantsFromAssemblyRegions(
+            final AuthHolder authHolder,
+            final SAMFileHeader header,
+            final Broadcast<ReferenceMultiSource> referenceBroadcast,
+            final Broadcast<HaplotypeCallerArgumentCollection> hcArgsBroadcast) {
+        return regionAndInterval -> {
+            final ReferenceMultiSourceToReferenceSequenceFileAdaptor referenceReader = new ReferenceMultiSourceToReferenceSequenceFileAdaptor(referenceBroadcast.getValue(), authHolder);
+            final HaplotypeCallerEngine hcEngine = new HaplotypeCallerEngine(hcArgsBroadcast.value(), header, referenceReader);
+            final List<VariantContext> variantContexts = hcEngine.callRegion(regionAndInterval._1(), new FeatureContext());
+            final SimpleInterval shardBoundary = regionAndInterval._2();
+            return variantContexts.stream().filter(shardBoundary::contains).collect(Collectors.toList());
+        };
+    }
+
+    /**
+     * WriteVariants, this is currently going to be horribly slow and explosive on a full size file since it performs a collect.
+     */
+    private void writeVariants(JavaRDD<VariantContext> variants) {
+        final List<VariantContext> collectedVariants = variants.collect();
+        final List<VariantContext> sortedVariants = collectedVariants.stream().sorted((o1, o2) -> IntervalUtils.compareLocatables(o1, o2, getReferenceSequenceDictionary())).collect(Collectors.toList());
```

Comment:
also add a comment here that it will be replaced by a shared write at some point (with a ticket number)